File: apps/android/app/src/main/java/com/litter/android/state/PetOverlayController.kt
package com.litter.android.state

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.provider.Settings
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import uniffi.codex_mobile_client.AppPetSummary
import uniffi.codex_mobile_client.AppSnapshotRecord
import uniffi.codex_mobile_client.AppServerTransportState
import uniffi.codex_mobile_client.ThreadSummaryStatus
import java.io.File

enum class PetAvatarState(val row: Int) {
    IDLE(0),
    RUNNING_RIGHT(1),
    RUNNING_LEFT(2),
    WAVING(3),
    JUMPING(4),
    FAILED(5),
    WAITING(6),
    RUNNING(7),
    REVIEW(8),
}

data class CachedPetPackage(
    val serverId: String,
    val id: String,
    val displayName: String,
    val spritesheetBytes: ByteArray,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is CachedPetPackage) return false
        return serverId == other.serverId &&
            id == other.id &&
            displayName == other.displayName &&
            spritesheetBytes.contentEquals(other.spritesheetBytes)
    }

    override fun hashCode(): Int {
        var result = serverId.hashCode()
        result = 31 * result + id.hashCode()
        result = 31 * result + displayName.hashCode()
        result = 31 * result + spritesheetBytes.contentHashCode()
        return result
    }
}

data class PetOverlayUiModel(
    val pet: CachedPetPackage,
    val state: PetAvatarState,
    val message: String?,
)

object PetOverlayController {
    private const val PREFS = "litter_pet_overlay"
    private const val KEY_VISIBLE = "visible"
    private const val KEY_OVERLAY_ENABLED = "overlay_enabled"
    private const val KEY_SERVER_ID = "server_id"
    private const val KEY_PET_ID = "pet_id"
    private const val KEY_PET_NAME = "pet_name"
    private const val KEY_DRAG_OFFSET_X = "drag_offset_x"
    private const val KEY_DRAG_OFFSET_Y = "drag_offset_y"
    private const val KEY_PET_SCALE = "pet_scale"
    private const val CACHE_DIR = "pets"

    const val MIN_PET_SCALE = 0.25f
    const val MAX_PET_SCALE = 5.0f
    const val DEFAULT_PET_SCALE = 1.0f

    var visible by mutableStateOf(false)
        private set
    var overlayEnabled by mutableStateOf(false)
        private set
    var selectedPet by mutableStateOf<CachedPetPackage?>(null)
        private set
    var isLoading by mutableStateOf(false)
        private set
    var errorMessage by mutableStateOf<String?>(null)
        private set
    var dragOffsetX by mutableFloatStateOf(24f)
        private set
    var dragOffsetY by mutableFloatStateOf(96f)
        private set
    var isDragging by mutableStateOf(false)
        private set
    var petScale by mutableFloatStateOf(DEFAULT_PET_SCALE)
        private set
    var isPinching by mutableStateOf(false)
        private set
    private var pinchInitialScale: Float = DEFAULT_PET_SCALE
    private var dragDirection by mutableStateOf(PetAvatarState.RUNNING_RIGHT)
    private var initialized = false

    fun initialize(context: Context) {
        if (initialized) return
        initialized = true
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        visible = prefs.getBoolean(KEY_VISIBLE, false)
        overlayEnabled = prefs.getBoolean(KEY_OVERLAY_ENABLED, false)
        dragOffsetX = prefs.getFloat(KEY_DRAG_OFFSET_X, 24f)
        dragOffsetY = prefs.getFloat(KEY_DRAG_OFFSET_Y, 96f)
        petScale = prefs.getFloat(KEY_PET_SCALE, DEFAULT_PET_SCALE)
            .coerceIn(MIN_PET_SCALE, MAX_PET_SCALE)
        val serverId = prefs.getString(KEY_SERVER_ID, null)
        val petId = prefs.getString(KEY_PET_ID, null)
        val name = prefs.getString(KEY_PET_NAME, null)
        if (!serverId.isNullOrBlank() && !petId.isNullOrBlank() && !name.isNullOrBlank()) {
            val file = cacheFile(context, serverId, petId)
            if (file.exists()) {
                selectedPet = CachedPetPackage(
                    serverId = serverId,
                    id = petId,
                    displayName = name,
                    spritesheetBytes = file.readBytes(),
                )
            }
        }
    }

    fun setVisible(context: Context, next: Boolean) {
        visible = next
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putBoolean(KEY_VISIBLE, next)
            .apply()
        syncOverlayService(context)
    }

    fun toggleVisible(context: Context) {
        setVisible(context, !visible)
    }

    fun setOverlayEnabled(context: Context, next: Boolean) {
        overlayEnabled = next
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putBoolean(KEY_OVERLAY_ENABLED, next)
            .apply()
        syncOverlayService(context)
    }

    suspend fun selectPet(context: Context, appModel: AppModel, serverId: String, pet: AppPetSummary) {
        isLoading = true
        errorMessage = null
        try {
            val packageResult = appModel.client.loadPet(serverId, pet.id)
            val cached = CachedPetPackage(
                serverId = serverId,
                id = packageResult.summary.id,
                displayName = packageResult.summary.displayName,
                spritesheetBytes = packageResult.spritesheetBytes,
            )
            cacheFile(context, serverId, cached.id).apply {
                parentFile?.mkdirs()
                writeBytes(cached.spritesheetBytes)
            }
            context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
                .edit()
                .putString(KEY_SERVER_ID, serverId)
                .putString(KEY_PET_ID, cached.id)
                .putString(KEY_PET_NAME, cached.displayName)
                .putBoolean(KEY_VISIBLE, true)
                .apply()
            selectedPet = cached
            visible = true
            syncOverlayService(context)
        } catch (error: Throwable) {
            errorMessage = error.message ?: "Unable to load pet."
        } finally {
            isLoading = false
        }
    }

    fun startDrag() {
        isDragging = true
    }

    fun dragBy(context: Context? = null, dx: Float, dy: Float) {
        dragOffsetX += dx
        dragOffsetY += dy
        if (dx > 0.5f) dragDirection = PetAvatarState.RUNNING_RIGHT
        if (dx < -0.5f) dragDirection = PetAvatarState.RUNNING_LEFT
        if (context != null) {
            persistPosition(context)
        }
    }

    fun setPosition(context: Context? = null, x: Float, y: Float) {
        val dx = x - dragOffsetX
        dragOffsetX = x
        dragOffsetY = y
        if (dx > 0.5f) dragDirection = PetAvatarState.RUNNING_RIGHT
        if (dx < -0.5f) dragDirection = PetAvatarState.RUNNING_LEFT
        if (context != null) {
            persistPosition(context)
        }
    }

    fun endDrag() {
        isDragging = false
    }

    fun startPinch() {
        pinchInitialScale = petScale
        isPinching = true
    }

    fun pinchBy(factor: Float) {
        if (!factor.isFinite() || factor <= 0f) return
        petScale = (pinchInitialScale * factor).coerceIn(MIN_PET_SCALE, MAX_PET_SCALE)
    }

    fun endPinch(context: Context) {
        isPinching = false
        persistScale(context)
    }

    fun setScale(context: Context, value: Float) {
        petScale = value.coerceIn(MIN_PET_SCALE, MAX_PET_SCALE)
        persistScale(context)
    }

    fun canDrawOverlays(context: Context): Boolean = Settings.canDrawOverlays(context)

    fun requestOverlayPermission(context: Context) {
        val intent = Intent(
            Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
            Uri.parse("package:${context.packageName}"),
        ).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        context.startActivity(intent)
    }

    fun shouldShowSystemOverlay(context: Context): Boolean =
        visible && overlayEnabled && selectedPet != null && canDrawOverlays(context)

    fun shouldShowInAppOverlay(context: Context): Boolean =
        visible && selectedPet != null && !shouldShowSystemOverlay(context)

    fun buildUiModel(snapshot: AppSnapshotRecord?): PetOverlayUiModel? {
        val pet = selectedPet ?: return null
        if (!visible) return null
        return PetOverlayUiModel(
            pet = pet,
            state = avatarState(snapshot),
            message = avatarMessage(snapshot),
        )
    }

    fun syncOverlayService(context: Context) {
        val appContext = context.applicationContext
        val intent = Intent(appContext, PetOverlayService::class.java).apply {
            action = PetOverlayService.ACTION_SYNC
        }
        if (shouldShowSystemOverlay(appContext)) {
            appContext.startService(intent)
        } else {
            appContext.stopService(intent)
        }
    }

    private fun persistPosition(context: Context) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putFloat(KEY_DRAG_OFFSET_X, dragOffsetX)
            .putFloat(KEY_DRAG_OFFSET_Y, dragOffsetY)
            .apply()
    }

    private fun persistScale(context: Context) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putFloat(KEY_PET_SCALE, petScale)
            .apply()
    }

    fun avatarState(snapshot: AppSnapshotRecord?): PetAvatarState {
        if (isLoading) return PetAvatarState.WAITING
        if (isDragging) return dragDirection
        if (snapshot == null) return PetAvatarState.IDLE
        if (snapshot.pendingApprovals.isNotEmpty() || snapshot.pendingUserInputs.isNotEmpty()) {
            return PetAvatarState.REVIEW
        }
        val activeKey = snapshot.activeThread
        val activeThread = activeKey?.let { key ->
            snapshot.threads.firstOrNull { it.key == key }
        }
        if (activeThread?.info?.status == ThreadSummaryStatus.SYSTEM_ERROR) {
            return PetAvatarState.FAILED
        }
        if (activeThread?.hasActiveTurn == true) {
            return PetAvatarState.RUNNING
        }
        if (snapshot.threads.any { it.hasActiveTurn }) {
            return PetAvatarState.RUNNING
        }
        if (snapshot.threads.any { it.info.status == ThreadSummaryStatus.SYSTEM_ERROR }) {
            return PetAvatarState.FAILED
        }
        val connected = snapshot.servers.any { it.transportState == AppServerTransportState.CONNECTED }
        return if (connected) PetAvatarState.IDLE else PetAvatarState.WAITING
    }

    fun avatarMessage(snapshot: AppSnapshotRecord?): String? {
        if (isLoading) return "Fetching pet..."
        if (isDragging) return null
        if (snapshot == null) return null
        if (snapshot.pendingApprovals.isNotEmpty()) return "Review needed"
        if (snapshot.pendingUserInputs.isNotEmpty()) return "Input needed"
        val activeKey = snapshot.activeThread
        val activeThread = activeKey?.let { key ->
            snapshot.threads.firstOrNull { it.key == key }
        }
        if (activeThread?.info?.status == ThreadSummaryStatus.SYSTEM_ERROR) return "Run failed"
        if (activeThread?.hasActiveTurn == true) return "Working..."
        if (snapshot.threads.any { it.hasActiveTurn }) return "Working..."
        if (snapshot.threads.any { it.info.status == ThreadSummaryStatus.SYSTEM_ERROR }) return "Thread failed"
        val connected = snapshot.servers.any { it.transportState == AppServerTransportState.CONNECTED }
        return if (connected) null else "Waiting for server"
    }

    private fun cacheFile(context: Context, serverId: String, petId: String): File {
        val safeServer = serverId.replace(Regex("[^A-Za-z0-9_.-]"), "_")
        val safePet = petId.replace(Regex("[^A-Za-z0-9_.-]"), "_")
        return File(File(context.filesDir, CACHE_DIR), "${safeServer}_${safePet}.webp")
    }
}

File: apps/ios/Sources/Litter/Models/PetOverlayController.swift
import Foundation
import Observation
import SwiftUI

enum PetAvatarState: Int {
    case idle = 0
    case runningRight = 1
    case runningLeft = 2
    case waving = 3
    case jumping = 4
    case failed = 5
    case waiting = 6
    case running = 7
    case review = 8
}

struct CachedPetPackage: Equatable {
    let serverId: String
    let id: String
    let displayName: String
    let spritesheetBytes: Data
}

@MainActor
@Observable
final class PetOverlayController {
    static let shared = PetOverlayController()

    static let minPetScale: CGFloat = 0.25
    static let maxPetScale: CGFloat = 5.0
    static let defaultPetScale: CGFloat = 1.0

    private let visibleKey = "litter.petOverlay.visible"
    private let serverIdKey = "litter.petOverlay.serverId"
    private let petIdKey = "litter.petOverlay.petId"
    private let petNameKey = "litter.petOverlay.petName"
    private let petScaleKey = "litter.petOverlay.petScale"

    private(set) var visible = false
    private(set) var selectedPet: CachedPetPackage?
    private(set) var isLoading = false
    private(set) var errorMessage: String?
    var dragOffset = CGSize(width: 24, height: 96)
    private(set) var isDragging = false
    private(set) var petScale: CGFloat = PetOverlayController.defaultPetScale
    private(set) var isPinching = false
    private var pinchInitialScale: CGFloat = PetOverlayController.defaultPetScale
    private var dragDirection = PetAvatarState.runningRight

    private init() {
        visible = UserDefaults.standard.bool(forKey: visibleKey)
        let savedScale = UserDefaults.standard.object(forKey: petScaleKey) as? Double
            ?? Double(Self.defaultPetScale)
        petScale = Self.clampScale(CGFloat(savedScale))
        guard let serverId = UserDefaults.standard.string(forKey: serverIdKey),
              let petId = UserDefaults.standard.string(forKey: petIdKey),
              let name = UserDefaults.standard.string(forKey: petNameKey),
              let data = try? Data(contentsOf: cacheURL(serverId: serverId, petId: petId))
        else { return }
        selectedPet = CachedPetPackage(
            serverId: serverId,
            id: petId,
            displayName: name,
            spritesheetBytes: data
        )
    }

    private static func clampScale(_ value: CGFloat) -> CGFloat {
        min(maxPetScale, max(minPetScale, value))
    }

    func setVisible(_ next: Bool) {
        visible = next
        UserDefaults.standard.set(next, forKey: visibleKey)
    }

    func selectPet(appModel: AppModel, serverId: String, pet: AppPetSummary) async {
        isLoading = true
        errorMessage = nil
        do {
            let package = try await appModel.client.loadPet(serverId: serverId, petId: pet.id)
            let cached = CachedPetPackage(
                serverId: serverId,
                id: package.summary.id,
                displayName: package.summary.displayName,
                spritesheetBytes: Data(package.spritesheetBytes)
            )
            try FileManager.default.createDirectory(
                at: cacheDirectory,
                withIntermediateDirectories: true
            )
            try cached.spritesheetBytes.write(to: cacheURL(serverId: serverId, petId: cached.id))
            UserDefaults.standard.set(serverId, forKey: serverIdKey)
            UserDefaults.standard.set(cached.id, forKey: petIdKey)
            UserDefaults.standard.set(cached.displayName, forKey: petNameKey)
            UserDefaults.standard.set(true, forKey: visibleKey)
            selectedPet = cached
            visible = true
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    func startDrag() {
        isDragging = true
    }

    func dragBy(_ translation: CGSize) {
        dragOffset.width += translation.width
        dragOffset.height += translation.height
        if translation.width > 0.5 { dragDirection = .runningRight }
        if translation.width < -0.5 { dragDirection = .runningLeft }
    }

    func endDrag() {
        isDragging = false
    }

    func startPinch() {
        guard !isPinching else { return }
        pinchInitialScale = petScale
        isPinching = true
    }

    func pinchBy(_ factor: CGFloat) {
        guard factor.isFinite, factor > 0 else { return }
        petScale = Self.clampScale(pinchInitialScale * factor)
    }

    func endPinch() {
        isPinching = false
        UserDefaults.standard.set(Double(petScale), forKey: petScaleKey)
    }

    func setScale(_ value: CGFloat) {
        petScale = Self.clampScale(value)
        UserDefaults.standard.set(Double(petScale), forKey: petScaleKey)
    }

    func avatarState(snapshot: AppSnapshotRecord?) -> PetAvatarState {
        if isLoading { return .waiting }
        if isDragging { return dragDirection }
        guard let snapshot else { return .idle }
        if !snapshot.pendingApprovals.isEmpty || !snapshot.pendingUserInputs.isEmpty {
            return .review
        }
        let activeThread = snapshot.activeThread.flatMap { key in
            snapshot.threads.first(where: { $0.key == key })
        }
        if activeThread?.info.status == .systemError { return .failed }
        if activeThread?.hasActiveTurn == true { return .running }
        if snapshot.threads.contains(where: \.hasActiveTurn) {
            return .running
        }
        if snapshot.threads.contains(where: { $0.info.status == .systemError }) {
            return .failed
        }
        return snapshot.servers.contains(where: \.isConnected) ? .idle : .waiting
    }

    func avatarMessage(snapshot: AppSnapshotRecord?) -> String? {
        if isLoading { return "Fetching pet..." }
        if isDragging { return nil }
        guard let snapshot else { return nil }
        if !snapshot.pendingApprovals.isEmpty { return "Review needed" }
        if !snapshot.pendingUserInputs.isEmpty { return "Input needed" }
        let activeThread = snapshot.activeThread.flatMap { key in
            snapshot.threads.first(where: { $0.key == key })
        }
        if activeThread?.info.status == .systemError { return "Run failed" }
        if activeThread?.hasActiveTurn == true { return "Working..." }
        if snapshot.threads.contains(where: \.hasActiveTurn) {
            return "Working..."
        }
        if snapshot.threads.contains(where: { $0.info.status == .systemError }) {
            return "Thread failed"
        }
        return snapshot.servers.contains(where: \.isConnected) ? nil : "Waiting for server"
    }

    private var cacheDirectory: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        return base.appendingPathComponent("Pets", isDirectory: true)
    }

    private func cacheURL(serverId: String, petId: String) -> URL {
        cacheDirectory.appendingPathComponent("\(safe(serverId))_\(safe(petId)).webp")
    }

    private func safe(_ value: String) -> String {
        value.map { char in
            char.isLetter || char.isNumber || char == "-" || char == "_" || char == "." ? char : "_"
        }.map(String.init).joined()
    }
}
